src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,321 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from datetime import datetime, timedelta
+from typing import Any, Generic, Sequence, TypeVar, Union
+
+import numpy as np
+
+T = TypeVar("T")
+
+Container = Union[list, np.ndarray]
+
+
+class RingBuffer(Generic[T]):
+    """A ring buffer with a fixed size.
+
+    Should work with most backends, tested with list and np.ndarrays.
+    """
+
+    class NoElements(Exception):
+        """Exception class, thrown when no elements are available when calling pop()."""
+
+    def __init__(self, container: Container) -> None:
+        """Initialize the ring buffer with the given container.
+
+        Args:
+            container: Container to store the data in.
+        """
+        self._container = container
+        self._write_index = 0
+        self._read_index = 0
+        self._len = 0
+
+    def __len__(self) -> int:
+        """Get current amount of elements.
+
+        Returns:
+            the amount of items that this container currently holds.
+        """
+        return self._len
+
+    @property
+    def maxlen(self) -> int:
+        """Get the max length.
+
+        Returns:
+            the max amount of items this container can hold.
+        """
+        return len(self._container)
+
+    def push(self, value: T) -> int:
+        """Push a new value into the ring buffer.
+
+        Args:
+            value: value to push into the ring buffer.
+
+        Returns:
+            the index in the ringbuffer.
+        """
+        if self._len == len(self._container):
+            # Move read position one forward, dropping the oldest written value
+            self._read_index = self._wrap(self._read_index + 1)
+        else:
+            self._len += 1
+
+        self._container[self._write_index] = value
+        value_index = self._write_index
+        self._write_index = self._wrap(self._write_index + 1)
+
+        return value_index
+
+    def pop(self) -> T:
+        """Remove the oldest value from the ring buffer and return it.
+
+        Raises:
+            NoElements: when no elements exist to pop.
+
+        Returns:
+            Oldest value found in the ring buffer.
+        """
+        if self._len == 0:
+            raise RingBuffer.NoElements()
+
+        val = self._container[self._read_index]
+        self._read_index = (self._read_index + 1) % len(self._container)
+        self._len -= 1
+
+        return val
+
+    @property
+    def full(self) -> bool:
```

Comment:
We try to stick to using something that looks like a question to name stuff that is a bool. What you had before was perfect: `is_full`.